Mark enumerator as complete if applicant did not answer

carolinedanzi opened the pull request:
### Description
If an applicant chose not to add any enumerator entities, we still showed them the question as if they had not completed it. This fixes this issue by ensuring the path exists in `ApplicantData` even if the applicant has no data to write for an enumerator

### Checklist
- [ ] Created tests which fail without the change (if possible)
- [ ] Extended the README / documentation, if necessary

### Issue(s)
Fixes #1141

hmhsu commented:
Could we do something like this https://github.com/seattle-uat/civiform/pull/1157? I really think this might be cleaner, but I fear I am missing some important pieces.

kendrickt reviewed (approved):
I removed the hack I put into the `ApplicantProgramBlocksController` and fixed the test.

This looks great to me. Two comments:

1. would make sense for `readRepeatedEntities` to return an Optional since that's what the other read methods return?
2. Using `putArrayIfMissing` is fine because we have lots of structure in the order of updates, but this reminds me of https://github.com/seattle-uat/civiform/issues/624. This won't work if the update is something like `applicant.household_members[3].job -> []`, which can't really happen, but I think `ApplicantData` should allow it.

carolinedanzi commented:
> Could we do something like this #1157? I really think this might be cleaner, but I fear I am missing some important pieces.

Oh neat, that's a nice improvement! I think you're right that something similar would work - I updated to use `putRepeatedEntities` instead with an empty list

carolinedanzi commented:
> I removed the hack I put into the `ApplicantProgramBlocksController` and fixed the test.
> 
> This looks great to me. Two comments:
> 
> 1. would make sense for `readRepeatedEntities` to return an Optional since that's what the other read methods return?
> 2. Using `putArrayIfMissing` is fine because we have lots of structure in the order of updates, but this reminds me of #624. This won't work if the update is something like `applicant.household_members[3].job -> []`, which can't really happen, but I think `ApplicantData` should allow it.

Thanks Kendrick! For #1, I think it makes more sense for `readRepeatedEntities` to return an empty list as it currently does - the optionals were due to other types not having a good empty/null default value. For lists, an empty list makes more sense

I'm not sure I understand #2, but I updated to use `putRepeatedEntities` with an empty list, is that better?

kendrickt commented on universal-application-tool-0.0.1/app/services/applicant/ApplicantServiceImpl.java:
```diff
@@ -258,6 +258,15 @@ private void stageEnumeratorUpdates(
+
+    // The applicant has seen this question, but has not supplied any entities.
+    // We need to still write this path so we can tell they have seen this question.
+    if (updates.isEmpty()) {
```
(blocking) I'd rather see `if !applicantData.hasPath(enumeratorPath) { put empty array; }

carolinedanzi replied:
Good point! Done
